### PR TITLE
fix(pbts): block Time must be Canonical

### PR DESCRIPTION
**Pull request by cason**
Motivated by [this comment](https://github.com/cometbft/cometbft/pull/2205#discussion_r1491321877).

---

#### PR checklist

- [x] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: internal/state/validation.go [hunk 1]
```diff
@@ -112,6 +113,10 @@ func validateBlock(state State, block *types.Block) error {
 	}
 
 	// Validate block Time
+	if block.Time != cmttime.Canonical(block.Time) {
```
Comment (sergio-mena) [inline internal/state/validation.go:116]: Is this check a sanity check?

**Review by sergio-mena — APPROVED** [review]
🚀

File: internal/state/validation.go (hunk 1 above)
Comment (cason) [reply internal/state/validation.go:116]: Yes, because of the "stripping", removing the monotonic part.

**Comment by sergio-mena** [comment]
@cason I found a solution. Check the two upcoming commits

File: internal/state/helpers_test.go [hunk 2]
```diff
@@ -259,3 +260,7 @@ func (app *testApp) ProcessProposal(
 	}
 	return &abci.ProcessProposalResponse{Status: abci.PROCESS_PROPOSAL_STATUS_ACCEPT}, nil
 }
+
+func makeStateNilParams(nVals, height int, chainID string) (sm.State, dbm.DB, map[string]types.PrivValidator) {
```
Comment (cason) [inline internal/state/helpers_test.go:264]: Should it be `DefaultParams`?

File: internal/state/state.go [hunk 3]
```diff
@@ -331,3 +333,46 @@ func MakeGenesisState(genDoc *types.GenesisDoc) (State, error) {
 		AppHash: genDoc.AppHash,
 	}, nil
 }
+
+// This is a testing function. Had to be placed here to avoid import cycles.
+func MakeState(nVals, height int, params *types.ConsensusParams, chainID string) (State, dbm.DB, map[string]types.PrivValidator) {
```
Comment (cason) [inline internal/state/state.go:338]: Not way to put this on a `_test.go` file? The problem is importing, right?

**Review by cason — COMMENTED** [review]
This works but it is not really nice...

File: internal/consensus/common_test.go [hunk 4]
```diff
@@ -1057,26 +1056,14 @@ func signDataIsEqual(v1 *types.Vote, v2 *cmtproto.Vote) bool {
...
-		valAddr := pk.PubKey().Address()
-		vals[i] = types.GenesisValidator{
-			Address: valAddr,
-			PubKey:  pk.PubKey(),
-			Power:   1000,
-			Name:    fmt.Sprintf("test%d", i),
-		}
-		privVals[valAddr.String()] = types.NewMockPVWithParams(pk, false, false)
-	}
+func makeStateNilParams(nVals int, chainID string) (sm.State, map[string]types.PrivValidator) {
```
Comment (lasarojc) [inline internal/consensus/common_test.go:1059]: To reduce the changes, call this one `makeState` and teh other one `makeStateWithCP`.

**Review by lasarojc — APPROVED** [review]
Just a suggestion to reduce the amount of changes.

File: internal/consensus/common_test.go (hunk 4 above)
Comment (cason) [reply internal/consensus/common_test.go:1059]: Yes, it is possible. I didn't want to undo all renaming Sergio did, but your comment makes sense.

File: internal/consensus/wal_test.go [hunk 5]
```diff
@@ -360,6 +362,28 @@ func TestWALPeriodicSync(t *testing.T) {
 	}
 }
 
+// FIXME: this helper is very similar to the one in internal/state/helpers_test.go.
```
Comment (cason) [inline internal/consensus/wal_test.go:365]: Moved it here, since it is not used anywhere else.

**Review by sergio-mena — APPROVED** [review]
👍

**Pull request merged**